pkg/kubelet/kubelet.go

```diff
@@ -1025,12 +1014,102 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 	if err := os.Mkdir(kl.getPodPluginsDir(uid), 0750); err != nil && !os.IsExist(err) {
 		return err
 	}
+	return nil
+}
 
-	// Make sure we have a pod infra container
-	var podInfraContainerID dockertools.DockerID
+func (kl *Kubelet) shouldContainerBeRestarted(pod *api.BoundPod, containerName, dockerContainerName string, uid types.UID) bool {
+	// Check RestartPolicy for dead container
+	recentContainers, err := dockertools.GetRecentDockerContainersWithNameAndUUID(kl.dockerClient, GetPodFullName(pod), uid, containerName)
+	if err != nil {
+		glog.Errorf("Error listing recent containers:%s", dockerContainerName)
+		// TODO(dawnchen): error handling here?
+	}
+	// set dead containers to unready state
+	for _, c := range recentContainers {
+		kl.readiness.remove(c.ID)
+	}
+
+	if len(recentContainers) > 0 {
+		if pod.Spec.RestartPolicy.Never != nil {
+			glog.Infof("Already ran container with name %s, do nothing",
+				dockerContainerName)
+			return false
+
+		}
+		if pod.Spec.RestartPolicy.OnFailure != nil {
+			// Check the exit code of last run
+			if recentContainers[0].State.ExitCode == 0 {
+				glog.Infof("Already successfully ran container with name %s, do nothing",
+					dockerContainerName)
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (kl *Kubelet) getPodInfraContainer(podFullName string, uid types.UID,
+	dockerContainers dockertools.DockerContainers) (api.PodStatus, dockertools.DockerID, bool) {
 	if podInfraDockerContainer, found, _ := dockerContainers.FindPodContainer(podFullName, uid, dockertools.PodInfraContainerName); found {
-		podInfraContainerID = dockertools.DockerID(podInfraDockerContainer.ID)
-	} else {
+		podInfraContainerID := dockertools.DockerID(podInfraDockerContainer.ID)
+		// Somethings is running - we can get Status from the Docker.
+		podStatus, err := kl.GetPodStatus(podFullName, uid)
+		if err != nil {
+			glog.Errorf("Unable to get pod with name %q and uid %q info with error(%v)", podFullName, uid, err)
+		}
+		return podStatus, podInfraContainerID, true
+	}
+	return api.PodStatus{}, "", false
+}
+
+func (kl *Kubelet) pullImageAndRunContainer(pod *api.BoundPod, container *api.Container, podVolumes *volumeMap,
```

Comment:
Lets document this method (mainly worried about the return values).

We should be able to return error instead of the last bool too btw. Seems like we always log the error instead of returning it. I know the flow was like that before, but it should be small enough to change safely :)